### PR TITLE
implemented BTreeMap/Set conversions ...

### PR DESCRIPTION
**Pull request by przygienda**
added the lacking BTreeMap/Set conversions ... 

I don't know where to plug in the to_redis_args based on the HashSet/Map implementation. Somehow BTreeMap has more traits and the implementation doesn't catch ...

### CODE REVIEW

**Comment by badboy** [comment]
Please don't commit your IDE's project files.
Please add tests.
Test failure unrelated, I restarted it.

**Comment by przygienda** [comment]
OK ... 

Can you tell me 

a) WHERE the implementation for Hash for the to_redis is? I can't find it since I need to include one that catches BTreeMap 

b) which test do you want the stuff to go into ?

**Comment by badboy** [comment]
I'm not entirely sure what you are asking. Do you mean [the already existing implementations for `HashMap` and `HashSet`](https://github.com/mitsuhiko/redis-rs/blob/b2ab3686b4b5f898720cb39e0285d4416c42649f/src/types.rs#L775-L806)?

There are [basic tests for both implementations](https://github.com/mitsuhiko/redis-rs/blob/b2ab3686b4b5f898720cb39e0285d4416c42649f/tests/test_basic.rs#L221-L254). You would need two similar test functions for your new implementations.

**Comment by przygienda** [comment]
OK, extended significantly, now the whole thing deals with maps as well gracefully generating them from BTreeSet and so on for HMGET and similar things. Tests added. IDE removed. Pls review ...

File: .gitignore [hunk 1]
```diff
@@ -1,3 +1,4 @@
+.idea
```
Comment (badboy) [inline .gitignore]: This belongs into your global gitignore, not in a project-specific one.

File: src/types.rs [hunk 2]
```diff
@@ -558,6 +570,58 @@ impl<T: ToRedisArgs> ToRedisArgs for Option<T> {
     }
 }
 
+impl<T: ToRedisArgs + Hash + Eq> ToRedisArgs for HashSet<T> {
+    fn to_redis_args(&self) -> Vec<Vec<u8>> {
+        let mut rv : Vec<&T> = Vec::with_capacity(self.len());
+
+        rv.extend(self.iter());
+        ToRedisArgs::make_arg_vec_ref(rv.as_slice())
```
Comment (badboy) [inline src/types.rs]: This looks like we are allocating twice. Once for the Vector in here, once for the Vector in `make_arg_vec_ref`, which seems kinda wasteful.

File: src/types.rs [hunk 3]
```diff
@@ -558,6 +570,58 @@ impl<T: ToRedisArgs> ToRedisArgs for Option<T> {
...
+}
+
+/// this flattens BTreeMap into something that goes well with HMSET
+impl<T: ToRedisArgs + Hash + Eq + Ord,
+     V: ToRedisArgs> ToRedisArgs for BTreeMap<T,V> {
+    fn to_redis_args(&self) -> Vec<Vec<u8>> {
+        let mut rv = Vec::with_capacity(self.len());
+
+        rv.extend(self.keys()
+            .zip(self.values())
```
Comment (badboy) [inline src/types.rs]: Overly complicated to zip keys and values. That's what the default `into_iter` already does. Use that.

File: src/types.rs [hunk 4]
```diff
@@ -558,6 +570,58 @@ impl<T: ToRedisArgs> ToRedisArgs for Option<T> {
...
+            .into_iter()
+            .flat_map(|e| {
```
Comment (badboy) [inline src/types.rs]: Use value destructuring here: `|(key,value)|`. Makes the code below far easier to read.

File: src/types.rs [hunk 5]
```diff
@@ -558,6 +570,58 @@ impl<T: ToRedisArgs> ToRedisArgs for Option<T> {
...
+                // otherwise things like HMSET will simply NOT work
+                assert!(e.0.is_single_arg() &&
+                        e.1.is_single_arg());
+
+                vec![e.0.to_redis_args(),
+                               e.1.to_redis_args()].into_iter() } )
+            .collect::<Vec<_>>());
```
Comment (badboy) [inline src/types.rs]: Avoid allocation. `extend` already expects an iterator.

File: src/types.rs [hunk 6]
```diff
@@ -558,6 +570,58 @@ impl<T: ToRedisArgs> ToRedisArgs for Option<T> {
...
+        ToRedisArgs::make_arg_vec(rv.as_slice())
```
Comment (badboy) [inline src/types.rs]: `rv.as_slice()` is just `&rv[..]`. Here simply `&rv` also works and is more idiomatic (IMO)

File: src/types.rs [hunk 7]
```diff
@@ -558,6 +570,58 @@ impl<T: ToRedisArgs> ToRedisArgs for Option<T> {
...
+    fn is_single_arg(&self) -> bool {
+        self.len() <= 1
+    }
+}
+
+/// this flattens BTreeMap into something that goes well with HMSET
+impl<T: ToRedisArgs + Hash + Eq + Ord,
+     V: ToRedisArgs> ToRedisArgs for BTreeMap<T,V> {
+    fn to_redis_args(&self) -> Vec<Vec<u8>> {
+        let mut rv = Vec::with_capacity(self.len());
```
Comment (badboy) [inline src/types.rs]: We actually need `self.len()*2` (key & value)

File: src/types.rs (hunk 2 above)
Comment (badboy) [inline src/types.rs]: `make_arg_vec_ref` was only added for this new use case and in both instances we double-allocate. We can simply inline this functionality for now

File: src/types.rs [hunk 8]
```diff
@@ -437,6 +438,17 @@ pub trait ToRedisArgs: Sized {
         rv
     }
 
+    /// This only exists internally as a workaround for the lack of
+    /// specialization.
+    #[doc(hidden)]
+    fn make_arg_vec_ref(items: &[&Self]) -> Vec<Vec<u8>> {
+        let mut rv = vec![];
+        for item in items.iter() {
+            rv.extend(item.to_redis_args().into_iter());
```
Comment (badboy) [inline src/types.rs]: No need for an explicit `.into_iter()` here.

**Comment by przygienda** [comment]
thanks. will attend to it. Fighting some other issues right now with the package on top ... Once that settles will come back here ...

File: src/types.rs (hunk 7 above)
Comment (przygienda) [inline src/types.rs]: done

File: src/types.rs (hunk 3 above)
Comment (przygienda) [inline src/types.rs]: done, just cruft that accumulated while I was experimenting

File: src/types.rs (hunk 4 above)
Comment (przygienda) [inline src/types.rs]: done

File: src/types.rs (hunk 5 above)
Comment (przygienda) [inline src/types.rs]: done, cool stuff, didn't know ...

File: src/types.rs (hunk 8 above)
Comment (przygienda) [inline src/types.rs]: done

File: src/types.rs [hunk 9]
```diff
@@ -558,6 +570,57 @@ impl<T: ToRedisArgs> ToRedisArgs for Option<T> {
...
+/// this flattens BTreeMap into something that goes well with HMSET
+impl<T: ToRedisArgs + Hash + Eq + Ord,
+     V: ToRedisArgs> ToRedisArgs for BTreeMap<T,V> {
+    fn to_redis_args(&self) -> Vec<Vec<u8>> {
+        let mut rv = Vec::with_capacity(self.len()*2);
+
+        rv.extend(self
+            .into_iter()
+            .flat_map(|e| {
+                let (key,value) = e;
```
Comment (badboy) [inline src/types.rs]: Why not `|(key,value)|`?

File: .gitignore (hunk 1 above)
Comment (badboy) [inline .gitignore]: This still needs to be removed.

**Comment by przygienda** [comment]
normalized bool representation. The test should actually write _all_ types into Redis and read back since right now they assume somehow the right "value type" shows up ...

**Comment by badboy** [comment]
Uhm, @przygienda, the last commit looks like it is related to #106 so it should be kept separate from the *Map implementations in this PR.

**Comment by przygienda** [comment]
sigh, I have no time to keep track of multiple branches so it's here. I have 3-4 different projects I'm fixing stuff to get it to work ... So please let that one just go in because as master is now it's buggy.

**Comment by badboy** [comment]
If you'd take a look at that other issue you'd see that using the strings `true` or `false` are probably the wrong way to go.
Thus I can't let this sit here as well. However, I can rip out that commit manually before merging.

**Comment by przygienda** [comment]
what's the "other issue".  I commit "0" and "1" as [u8;1] which is what you had as "status". redis on read does return Value::Data(..) which is what I check while I left your checks in place ...

**Comment by badboy** [comment]
#106 

Oh sorry. At first glance I somehow overlooked the `"0"` and `"1"` and just assumed it was true/false. In that case it might be more likely to be correct. I'll still pull this apart from the initial issue

**Comment by przygienda** [comment]
cool, cherry pick or whatever you like ... It's just one diff ... thanks much ...

File: src/types.rs (hunk 6 above)
Comment (przygienda) [inline src/types.rs]: done

**Comment by przygienda** [comment]
@badboy: ok, beta fails due to unrelated stuff ... otherwise e'thing reworked. Added more bug-fixes while working with it ... Let me know what you need more ..

**Comment by badboy** [comment]
I'll hope to do a final review this week and then merge it.

**Comment by badboy** [comment]
You didn't address the double-allocation at all. Could you take another look?

File: .gitignore (hunk 1 above)
Comment (przygienda) [inline .gitignore]: done

**Comment by przygienda** [comment]
so what's still outstanding ? Took care of pretty much everything requested ...

**Comment by badboy** [comment]
My last final review. I got a bit swamped with other work the last two weeks. I'll take care of this tomorrow.

**Comment by przygienda** [comment]
Have it working now under pretty heavy load in my app, 10K+ resulting key writes and the only gotcha was the empty sets/maps that Redis does not support. I put a comment in ... running it over redis 3.2.6 up to 3.2.8 ... Thanks for the outstanding package ...

**Comment by badboy** [comment]
I reordered & rebased your commits. Merging in #119

**Pull request closed without merging**